File: fsharp/forth/Forth.fs
module Forth

let newCommands (commands: Map<string, list<int> -> option<list<int>>>) (commandTokens: list<string>) =
    match commandTokens with
    | key :: _ when (System.Int32.TryParse key) |> fst -> None
    | key :: rest ->

        let defaultFn = fun _ -> None

        let fn =
            (Some, rest)
            ||> List.fold (fun s t ->
                match System.Int32.TryParse t with
                | true, i -> s >> Option.map (fun xs -> i :: xs)
                | _, _ ->
                    let cmdFn =
                        (defaultFn, commands.TryFind t)
                        ||> Option.defaultValue

                    s >> Option.bind cmdFn)

        commands |> Map.add key fn |> Some
    | _ -> None

let rec eval (commands: Map<string, list<int> -> option<list<int>>>) (tokens: list<string>) (result: list<int>) =
    match tokens with
    | [] -> result |> List.rev |> Some
    | x :: xs ->
        match System.Int32.TryParse x with
        | true, i -> eval commands xs (i :: result)
        | _, _ ->
            match commands.TryFind x with
            | Some f -> result |> f |> Option.bind (eval commands xs)
            | _ ->
                match x with
                | ":" ->
                    let commandTokens = xs |> List.takeWhile (fun t -> t <> ";")

                    (commands, commandTokens)
                    ||> newCommands
                    |> Option.bind (fun commands' ->
                        let newTokens =
                            xs
                            |> List.skipWhile (fun t -> t <> ";")
                            |> List.skip 1

                        eval commands' newTokens result)
                | _ -> None

let unary (f: int -> option<list<int>>) (xs: list<int>) =
    match xs with
    | x :: xs -> (f x) |> Option.map (fun r -> r @ xs)
    | _ -> None

let binary (f: int -> int -> option<list<int>>) (xs: list<int>) =
    match xs with
    | b :: a :: xs -> (f a b) |> Option.map (fun r -> r @ xs)
    | _ -> None

let evaluate (xs: list<string>) : option<list<int>> =
    let tokens =
        xs
        |> String.concat " "
        |> fun s ->
            s.ToUpperInvariant().Split [| ' ' |]
            |> Array.toList

    let commands =
        Map [ ("+", binary (fun a b -> Some [ a + b ]))
              ("-", binary (fun a b -> Some [ a - b ]))
              ("*", binary (fun a b -> Some [ a * b ]))
              ("/", binary (fun a b -> if b = 0 then None else Some [ a / b ]))
              ("DUP", unary (fun x -> Some [ x; x ]))
              ("DROP", unary (fun _ -> Some []))
              ("SWAP", binary (fun a b -> Some [ a; b ]))
              ("OVER", binary (fun a b -> Some [ a; b; a ])) ]

    eval commands tokens []

File: csharp/forth/Forth.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Operator
{
    Add, Substract, Multiply, Divide
}
public static class Forth
{
    public static string Evaluate(string[] instructions)
    {
        var list = new List<int>();
        IDictionary<string, string> userInputs = new Dictionary<string, string>();
        foreach (var instruction in instructions)
        {
            if (instruction.Contains(':'))
                userInputs = UserDefinedWords(instruction.ToLowerInvariant(), userInputs);
            else
            {
                string[] split = ReplaceInstuction(userInputs, instruction.ToLowerInvariant()).Split(' ', StringSplitOptions.RemoveEmptyEntries);
                foreach (var item in split)
                {
                    list = item switch
                    {
                        "+" => ArithmeticOperatio(list, Operator.Add),
                        "-" => ArithmeticOperatio(list, Operator.Substract),
                        "*" => ArithmeticOperatio(list, Operator.Multiply),
                        "/" => ArithmeticOperatio(list, Operator.Divide),
                        "dup" => list.Any() ? new List<int>(list) { list.Last() } : throw new InvalidOperationException(),
                        "drop" => list.Any() ? list.SkipLast(1).ToList() : throw new InvalidOperationException(),
                        "swap" => Swap(list).ToList(),
                        "over" => Over(list).ToList(),
                        _ when int.TryParse(item, out int value) => new List<int>(list) { value },
                        _ => throw new InvalidOperationException()
                    };
                }
            }
        }

        return string.Join(' ', list);
    }

    private static string ReplaceInstuction(IDictionary<string, string> inputs, string instruction)
    {
        foreach (var (key, value) in inputs)
            instruction = instruction.Replace(key, value);
        return instruction;
    }

    private static IDictionary<string, string> UserDefinedWords(string instruction, IDictionary<string, string> commands)
    {
        var split = instruction.Split(new char[] { ':', ';', ' ' }, StringSplitOptions.RemoveEmptyEntries);
        string key = string.Empty, value = string.Empty;
        for (int i = 0; i < split.Length; i++)
        {

            if (i == 0)
                key = int.TryParse(split[i], out int result) ? throw new InvalidOperationException() : split[i];
            else
                value += commands.ContainsKey(split[i]) ? $"{commands[split[i]]} " : $"{split[i]} ";
        }
        if (commands.TryAdd(key, value))
            return commands;
        commands[key] = value;
        return commands;
    }

    private static IEnumerable<int> Swap(IEnumerable<int> numbers) => (numbers.Count() < 2) ? throw new InvalidOperationException() : numbers.Take(numbers.Count() - 2).Concat(numbers.TakeLast(2).Reverse());

    private static IEnumerable<int> Over(IEnumerable<int> numbers) => (numbers.Count() < 2) ? throw new InvalidOperationException() : numbers.Append(numbers.ElementAt(numbers.Count() - 2));

    private static List<int> ArithmeticOperatio(IEnumerable<int> numbers, Operator op)
    {
        if (numbers.Count() < 2)
            throw new InvalidOperationException();
        var result = numbers.Aggregate((acc, cur) => {
            return op switch
            {
                Operator.Add => acc + cur,
                Operator.Substract => acc - cur,
                Operator.Multiply => acc * cur,
                Operator.Divide => acc / cur,
                _ => throw new InvalidOperationException()
            };
        });

        return new[] { result }.ToList();
    }

}